Match transactions by their TID and XXRequest to XXResponse

triller-telekom opened the pull request:
Fixes #1114

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

cdjackson commented:
Thanks Stefan.

I guess the big question - how well tested is this? :)

I'll run some tests, but do you consider this is ready for "prime time" yet?

triller-telekom commented:
I will create an intermediate version that I can give to our testers who have different devices and who could test this too.

But so far, I am really surprised on how good it works. No exceptions, no "feature" that did not work, all was working as expected.

If you have some corner cases that should be tested, let me know. Maybe we can also test them somehow.

cdjackson commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ActiveEndpointsRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/IeeeAddressRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NetworkAddressRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NodeDescriptorRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLeaveRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UserDescriptorRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/SimpleDescriptorRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ComplexDescriptorRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/PowerDescriptorRequest.java  -1
         

Clones removed
==============
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=6068786&bid=19873994)

triller-telekom commented:
As mentioned in the comment https://github.com/zsmartsystems/com.zsmartsystems.zigbee/issues/1114#issuecomment-680809391

I have written some code to address the broadcast issue in a separate commit.

So far this is untested though.

cdjackson commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementPermitJoiningRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/PowerDescriptorRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementBindRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/SimpleDescriptorRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NodeDescriptorRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ComplexDescriptorRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/IeeeAddressRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UserDescriptorRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UnbindRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ActiveEndpointsRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NetworkAddressRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/BindRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLeaveRequest.java  2
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  2
         

Clones removed
==============
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=6068786&bid=19873994)

triller-telekom commented:
I am still debugging it, because apparently we can also get `NetworkAddressResposnse`s from the coordinator to our unicast `NetworkaddressRequest` which doesn't match with the current implementation, hence I have set the PR to WIP.

cdjackson commented:
What do you mean? Can you provide a log? I wouldn't have thought that the responses from the coordinator would change. You will get a `NetworkAddressResponses` from the coordinator under some circumstances.

triller-telekom commented:
What I mean is this response:

```
2020-08-27T15:47:12.689+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
```

in

```
2020-08-27T15:47:12.109+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisDeviceJoinedNetworkEvent [networkAddress=47296, ieeeAddress=00124B000EE99CDB, parentAddress=0]
2020-08-27T15:47:12.119+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: nodeStatusUpdate - node status is UNSECURED_JOIN, network address is B8C0.
2020-08-27T15:47:12.139+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNetworkDiscoverer] - 00124B000EE99CDB: Device status updated. NWK=B8C0
2020-08-27T15:47:12.139+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNetworkDiscoverer] - 00124B000EE99CDB: NWK Discovery add node B8C0
2020-08-27T15:47:12.149+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNode] - 00124B000EE99CDB: Node state updated from UNKNOWN to ONLINE
2020-08-27T15:47:12.149+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: Updating node NWK=B8C0
2020-08-27T15:47:12.159+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.database.ZigBeeNetworkDatabaseManager] - 00124B000EE99CDB: Data store: Deferring write for 250ms.
2020-08-27T15:47:12.169+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeDiscoveryExtension] - 00124B000EE99CDB: DISCOVERY Extension: Adding discoverer for node
2020-08-27T15:47:12.169+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: created discoverer
2020-08-27T15:47:12.189+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: start discovery
2020-08-27T15:47:12.189+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: starting new tasks [NWK_ADDRESS, NODE_DESCRIPTOR, ACTIVE_ENDPOINTS, POWER_DESCRIPTOR]
2020-08-27T15:47:12.199+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: scheduled [NWK_ADDRESS, NODE_DESCRIPTOR, ACTIVE_ENDPOINTS, POWER_DESCRIPTOR]
2020-08-27T15:47:12.229+0200   212  DEBUG          [symbolicName=or.openhab.binding.zigbee.core, className=or.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - 00124B000EE99CDB: Node discovery not complete
2020-08-27T15:47:12.289+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: running NWK_ADDRESS
2020-08-27T15:47:12.299+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - 00124B000EE99CDB: Creating new Transaction Queue
2020-08-27T15:47:12.299+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: Set profile to ZigBeeTransactionProfile [maxOutstandingTransactions=2, interTransactionDelay=50, maxRetries=2]
2020-08-27T15:47:12.309+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: Added transaction to queue, len=1, transaction=ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=0, state=WAITING, sendCnt=0, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=--, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:12.309+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=1, sleepy=0/3
2020-08-27T15:47:12.319+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - B8C0/0: Sending ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=0, state=WAITING, sendCnt=0, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:12.319+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerAdded: 1 outstanding
2020-08-27T15:47:12.329+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX CMD: NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]
2020-08-27T15:47:12.339+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=B8C0/0, profile=0000, cluster=0000, addressMode=DEVICE, radius=8, apsSecurity=false, ackRequest=true, apsCounter=06, rssi=--, lqi=--, payload=05 DB 9C E9 0E 00 4B 12 00 00 00]
2020-08-27T15:47:12.339+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.ZigBeeDongleTelegesis] - Telegesis send: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=null]
2020-08-27T15:47:12.349+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis queue: 1
2020-08-27T15:47:12.349+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=null]
2020-08-27T15:47:12.409+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis Data:AT+SENDUCASTB:0B,B8C0,00,00,0000,0000
??? K   

2020-08-27T15:47:12.419+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.database.ZigBeeNetworkDatabaseManager] - 00124B000EE99CDB: Data store: Writing node.
2020-08-27T15:47:12.459+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:SEQ:0B
2020-08-27T15:47:12.499+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:OK
2020-08-27T15:47:12.499+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=11, status=SUCCESS]
2020-08-27T15:47:12.499+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX STA: msgTag=05 state=TX_ACK
2020-08-27T15:47:12.499+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionProgress: TID=05, state=TX_ACK, outstanding=1
2020-08-27T15:47:12.509+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction state changed: nwk=B8C0, TID=05, event=TX_ACK, state=TRANSMITTED
2020-08-27T15:47:12.509+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:ACK:0B
2020-08-27T15:47:12.509+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisAckMessageEvent [messageId=11]
2020-08-27T15:47:12.559+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX STA: msgTag=05 state=RX_ACK
2020-08-27T15:47:12.559+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionProgress: TID=05, state=RX_ACK, outstanding=1
2020-08-27T15:47:12.559+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction state changed: nwk=B8C0, TID=05, event=RX_ACK, state=ACKED
2020-08-27T15:47:12.609+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:RX:0000,0000,00,00,8000,0C: ??? K ??,00,FF
2020-08-27T15:47:12.639+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisReceiveMessageEvent [ieeeAddress=null, networkAddress=0, profileId=0, destinationEp=0, sourceEp=0, clusterId=32768, messageData=05 00 DB 9C E9 0E 00 4B 12 00 C0 B8, rssi=0, lqi=255]
2020-08-27T15:47:12.669+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=0000/0, profile=0000, cluster=8000, addressMode=null, radius=0, apsSecurity=false, ackRequest=false, apsCounter=--, rssi=--, lqi=--, payload=05 00 DB 9C E9 0E 00 4B 12 00 C0 B8]
2020-08-27T15:47:12.689+0200   212  DEBUG          [symbolicName=or.openhab.binding.zigbee.core, className=or.openhab.binding.zigbee.internal.ZigBeeDataStore] - 00124B000EE99CDB: ZigBee saving network state complete.
2020-08-27T15:47:12.689+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:12.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionCommand: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]] 
2020-08-27T15:47:12.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: destinationAddress=47296 isBroadcast=false
2020-08-27T15:47:12.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan sourceAddress=0000/0
2020-08-27T15:47:12.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Not broadcast
2020-08-27T15:47:12.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Addresses dont match
2020-08-27T15:47:12.729+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Response matches: command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0] response=NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:12.889+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:RX:B8C0,0000,00,00,0013,0C: ????? K ?,-32,FF
2020-08-27T15:47:12.889+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisReceiveMessageEvent [ieeeAddress=null, networkAddress=47296, profileId=0, destinationEp=0, sourceEp=0, clusterId=19, messageData=00 C0 B8 DB 9C E9 0E 00 4B 12 00 80, rssi=-50, lqi=255]
2020-08-27T15:47:12.899+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX APS: ZigBeeApsFrame [sourceAddress=B8C0/0, destinationAddress=0000/0, profile=0000, cluster=0013, addressMode=null, radius=0, apsSecurity=false, ackRequest=false, apsCounter=--, rssi=--, lqi=--, payload=00 C0 B8 DB 9C E9 0E 00 4B 12 00 80]
2020-08-27T15:47:12.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: DeviceAnnounce [B8C0/0 -> 0000/0, cluster=0013, TID=00, nwkAddrOfInterest=B8C0, ieeeAddr=00124B000EE99CDB, capability=80]
2020-08-27T15:47:12.919+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionCommand: DeviceAnnounce [B8C0/0 -> 0000/0, cluster=0013, TID=00, nwkAddrOfInterest=B8C0, ieeeAddr=00124B000EE99CDB, capability=80] 
2020-08-27T15:47:12.919+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNetworkDiscoverer] - 00124B000EE99CDB: Device announce received. From B8C0, for B8C0
2020-08-27T15:47:12.929+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNetworkDiscoverer] - 00124B000EE99CDB: NWK Discovery add node B8C0
2020-08-27T15:47:12.929+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNode] - 00124B000EE99CDB: Node state updated from UNKNOWN to ONLINE
2020-08-27T15:47:12.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: Updating node NWK=B8C0
2020-08-27T15:47:12.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: Node update. NWK Address=B8C0
2020-08-27T15:47:12.949+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: Node B8C0 is not updated
2020-08-27T15:47:13.079+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:RX:B8C0,0104,03,01,0500,07:  ,-32,FF
2020-08-27T15:47:13.089+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisReceiveMessageEvent [ieeeAddress=null, networkAddress=47296, profileId=260, destinationEp=3, sourceEp=1, clusterId=1280, messageData=19 00 01 15 00 01 10, rssi=-50, lqi=255]
2020-08-27T15:47:13.099+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX APS: ZigBeeApsFrame [sourceAddress=B8C0/1, destinationAddress=0000/3, profile=0104, cluster=0500, addressMode=null, radius=0, apsSecurity=false, ackRequest=false, apsCounter=--, rssi=--, lqi=--, payload=19 00 01 15 00 01 10]
2020-08-27T15:47:13.129+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX ZCL: ZclHeader [frameType=CLUSTER_SPECIFIC_COMMAND, manufacturerSpecific=false, direction=SERVER_TO_CLIENT, disableDefaultResponse=true, manufacturerCode=0, sequenceNumber=0, commandId=1]
2020-08-27T15:47:13.149+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - 00124B000EE99CDB: Endpoint 1. Unknown remote endpoint
2020-08-27T15:47:13.149+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - Incoming message from node B8C0 did not translate to command
2020-08-27T15:47:13.769+0200   185  DEBUG          [symbolicName=org.eclipse.smarthome.binding.homematic, className=org.eclipse.smarthome.binding.homematic.internal.communicator.AbstractHomematicGateway] - Received new (String) value 'internal' for 'CENTRAL:0#PONG' from gateway with id 'internal'
2020-08-27T15:47:13.839+0200   135  DEBUG          [symbolicName=org.eclipse.smarthome.core, className=org.eclipse.smarthome.core.items.events.ItemStateEvent] - homematic_bridge_internal_DUTY_CYCLE_RATIO updated to 2
2020-08-27T15:47:20.519+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction complete: ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=8210, state=FAILED, sendCnt=1, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:20.519+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerRemoved: 0 outstanding
2020-08-27T15:47:20.519+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: transactionComplete, state=FAILED, outstanding=0
2020-08-27T15:47:20.529+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: Added transaction to queue, len=1, transaction=ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=8210, state=WAITING, sendCnt=1, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:20.529+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=1, sleepy=0/3
2020-08-27T15:47:20.539+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - B8C0/0: Sending ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=8220, state=WAITING, sendCnt=1, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:20.539+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerAdded: 1 outstanding
2020-08-27T15:47:20.549+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX CMD: NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]
2020-08-27T15:47:20.559+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=B8C0/0, profile=0000, cluster=0000, addressMode=DEVICE, radius=8, apsSecurity=false, ackRequest=true, apsCounter=07, rssi=--, lqi=--, payload=05 DB 9C E9 0E 00 4B 12 00 00 00]
2020-08-27T15:47:20.559+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.ZigBeeDongleTelegesis] - Telegesis send: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=null]
2020-08-27T15:47:20.579+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis queue: 1
2020-08-27T15:47:20.579+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=null]
2020-08-27T15:47:20.639+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis Data:AT+SENDUCASTB:0B,B8C0,00,00,0000,0000
??? K   

2020-08-27T15:47:20.689+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:SEQ:0E
2020-08-27T15:47:20.699+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:OK
2020-08-27T15:47:20.699+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisSendUnicastCommand [address=47296, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=05 DB 9C E9 0E 00 4B 12 00 00 00, messageId=14, status=SUCCESS]
2020-08-27T15:47:20.709+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:ACK:0E
2020-08-27T15:47:20.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX STA: msgTag=05 state=TX_ACK
2020-08-27T15:47:20.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionProgress: TID=05, state=TX_ACK, outstanding=1
2020-08-27T15:47:20.709+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction state changed: nwk=B8C0, TID=05, event=TX_ACK, state=TRANSMITTED
2020-08-27T15:47:20.719+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisAckMessageEvent [messageId=14]
2020-08-27T15:47:20.729+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX STA: msgTag=05 state=RX_ACK
2020-08-27T15:47:20.729+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionProgress: TID=05, state=RX_ACK, outstanding=1
2020-08-27T15:47:20.729+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction state changed: nwk=B8C0, TID=05, event=RX_ACK, state=ACKED
2020-08-27T15:47:20.749+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:RX:0000,0000,00,00,8000,0C: ??? K ??,00,FF
2020-08-27T15:47:20.759+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisReceiveMessageEvent [ieeeAddress=null, networkAddress=0, profileId=0, destinationEp=0, sourceEp=0, clusterId=32768, messageData=05 00 DB 9C E9 0E 00 4B 12 00 C0 B8, rssi=0, lqi=255]
2020-08-27T15:47:20.759+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=0000/0, profile=0000, cluster=8000, addressMode=null, radius=0, apsSecurity=false, ackRequest=false, apsCounter=--, rssi=--, lqi=--, payload=05 00 DB 9C E9 0E 00 4B 12 00 C0 B8]
2020-08-27T15:47:20.769+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:20.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionCommand: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]] 
2020-08-27T15:47:20.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: destinationAddress=47296 isBroadcast=false
2020-08-27T15:47:20.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan sourceAddress=0000/0
2020-08-27T15:47:20.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Not broadcast
2020-08-27T15:47:20.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Addresses dont match
2020-08-27T15:47:20.789+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Response matches: command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0] response=NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=05, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:26.439+0200   470  DEBUG          [symbolicName=de.telekom.smarthomeb2c.core, className=de.telekom.smarthome.b2cmodul.services.backup.BackupProviderImpl] - running plugin periodic configuration backup, queue size is: 0
2020-08-27T15:47:28.719+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction complete: ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=16410, state=FAILED, sendCnt=2, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:28.719+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerRemoved: 0 outstanding
2020-08-27T15:47:28.719+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: transactionComplete, state=FAILED, outstanding=0
2020-08-27T15:47:28.719+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: transactionComplete exceeded max retries 2
2020-08-27T15:47:28.729+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction cancelled: ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=16410, state=CANCELLED, sendCnt=2, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:28.739+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction complete: ZigBeeTransaction [ieeeAddress=00124B000EE99CDB queueTime=16430, state=CANCELLED, sendCnt=2, command=NetworkAddressRequest [0000/0 -> B8C0/0, cluster=0000, TID=05, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:28.739+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: NetworkAddressRequest returned null
2020-08-27T15:47:28.739+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerRemoved: 0 outstanding
2020-08-27T15:47:28.749+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - 00124B000EE99CDB: transactionComplete but not outstanding, state=CANCELLED, outstanding=0
2020-08-27T15:47:28.749+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=0, sleepy=0/3
2020-08-27T15:47:28.749+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=0, sleepy=0/3
2020-08-27T15:47:28.749+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - Broadcast: Added transaction to queue, len=1, transaction=ZigBeeTransaction [ieeeAddress=null queueTime=0, state=WAITING, sendCnt=0, command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=--, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:28.759+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=1, sleepy=0/3
2020-08-27T15:47:28.759+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - FFFF/0: Sending ZigBeeTransaction [ieeeAddress=null queueTime=10, state=WAITING, sendCnt=0, command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:28.759+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerAdded: 1 outstanding
2020-08-27T15:47:28.769+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX CMD: NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]
2020-08-27T15:47:28.779+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=FFFF/0, profile=0000, cluster=0000, addressMode=DEVICE, radius=8, apsSecurity=false, ackRequest=true, apsCounter=08, rssi=--, lqi=--, payload=06 DB 9C E9 0E 00 4B 12 00 00 00]
2020-08-27T15:47:28.779+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.ZigBeeDongleTelegesis] - Telegesis send: TelegesisSendMulticastCommand [radius=8, address=65535, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=06 DB 9C E9 0E 00 4B 12 00 00 00]
2020-08-27T15:47:28.779+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis queue: 1
2020-08-27T15:47:28.789+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis: TelegesisSendMulticastCommand [radius=8, address=65535, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=06 DB 9C E9 0E 00 4B 12 00 00 00]
2020-08-27T15:47:28.839+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - TX Telegesis Data:AT+SENDMCASTB:0B,08,FFFF,00,00,0000,0000
??? K   

2020-08-27T15:47:28.879+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:OK
2020-08-27T15:47:28.879+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisSendMulticastCommand [radius=8, address=65535, sourceEp=0, destEp=0, profileId=0, clusterId=0, messageData=06 DB 9C E9 0E 00 4B 12 00 00 00, status=SUCCESS]
2020-08-27T15:47:28.889+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX STA: msgTag=06 state=TX_ACK
2020-08-27T15:47:28.889+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionProgress: TID=06, state=TX_ACK, outstanding=1
2020-08-27T15:47:28.899+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction state changed: nwk=FFFF, TID=06, event=TX_ACK, state=TRANSMITTED
2020-08-27T15:47:28.949+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis Data:RX:0000,0000,00,00,8000,0C: ??? K ??,00,FF
2020-08-27T15:47:28.959+0200   227  DEBUG          [symbolicName=com.zsmartsystems.zigbee.dongle.telegesis, className=com.zsmartsystems.zigbee.dongle.telegesis.internal.TelegesisFrameHandler] - RX Telegesis: TelegesisReceiveMessageEvent [ieeeAddress=null, networkAddress=0, profileId=0, destinationEp=0, sourceEp=0, clusterId=32768, messageData=06 00 DB 9C E9 0E 00 4B 12 00 C0 B8, rssi=0, lqi=255]
2020-08-27T15:47:28.979+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX APS: ZigBeeApsFrame [sourceAddress=0000/0, destinationAddress=0000/0, profile=0000, cluster=8000, addressMode=null, radius=0, apsSecurity=false, ackRequest=false, apsCounter=--, rssi=--, lqi=--, payload=06 00 DB 9C E9 0E 00 4B 12 00 C0 B8]
2020-08-27T15:47:28.979+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=06, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:28.999+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionCommand: NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=06, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]] 
2020-08-27T15:47:28.999+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: destinationAddress=65535 isBroadcast=true
2020-08-27T15:47:28.999+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan sourceAddress=0000/0
2020-08-27T15:47:29.019+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Stefan: Response matches: command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0] response=NetworkAddressResponse [0000/0 -> 0000/0, cluster=8000, TID=06, status=SUCCESS, ieeeAddrRemoteDev=00124B000EE99CDB, nwkAddrRemoteDev=B8C0, startIndex=null, nwkAddrAssocDevList=[]]
2020-08-27T15:47:29.159+0200   185  DEBUG          [symbolicName=org.eclipse.smarthome.binding.homematic, className=org.eclipse.smarthome.binding.homematic.internal.communicator.AbstractHomematicGateway] - Received new (String) value 'internal' for 'CENTRAL:0#PONG' from gateway with id 'internal'
2020-08-27T15:47:29.249+0200   135  DEBUG          [symbolicName=org.eclipse.smarthome.core, className=org.eclipse.smarthome.core.items.events.ItemStateEvent] - homematic_bridge_internal_DUTY_CYCLE_RATIO updated to 2
2020-08-27T15:47:36.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction complete: ZigBeeTransaction [ieeeAddress=null queueTime=8160, state=FAILED, sendCnt=1, command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:36.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerRemoved: 0 outstanding
2020-08-27T15:47:36.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - Broadcast: transactionComplete, state=FAILED, outstanding=0
2020-08-27T15:47:36.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - Broadcast: transactionComplete exceeded max retries 1
2020-08-27T15:47:36.909+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransaction] - Transaction cancelled: ZigBeeTransaction [ieeeAddress=null queueTime=8160, state=CANCELLED, sendCnt=1, command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:36.919+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: NetworkAddressRequest returned null
2020-08-27T15:47:36.919+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: NetworkAddressRequest failed after unicast and broadcast
2020-08-27T15:47:36.929+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction complete: ZigBeeTransaction [ieeeAddress=null queueTime=8160, state=CANCELLED, sendCnt=1, command=NetworkAddressRequest [0000/0 -> FFFF/0, cluster=0000, TID=06, ieeeAddr=00124B000EE99CDB, requestType=0, startIndex=0]]
2020-08-27T15:47:36.929+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: request NWK_ADDRESS failed. Retry 1, wait 2324ms before retry.
2020-08-27T15:47:36.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - transactionListenerRemoved: 0 outstanding
2020-08-27T15:47:36.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionQueue] - Broadcast: transactionComplete but not outstanding, state=CANCELLED, outstanding=0
2020-08-27T15:47:36.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=0, sleepy=0/3
2020-08-27T15:47:36.939+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - Transaction Manager: Send Next transaction. outstandingTransactions=0, outstandingQueues=0, sleepy=0/3
2020-08-27T15:47:39.259+0200   211  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.discovery.ZigBeeNodeServiceDiscoverer] - 00124B000EE99CDB: Node SVC Discovery: running NWK_ADDRESS
```

The "Stefan" logs come from my debug adjustments within `ZigbeeTransaction`:

```
public void commandReceived(ZigBeeCommand receivedCommand) {
        if (responseMatcher == null) {
            return;
        }

        // Ensure that received command is not processed before command is sent
        // and hence transaction ID for the command set.
        synchronized (command) {

            if (Objects.equals(command.getTransactionId(), receivedCommand.getTransactionId())) {
                int destinationAddress = ((ZdoRequest) command).getDestinationAddress().getAddress();
                logger.debug("Stefan: destinationAddress={} isBroadcast={}", destinationAddress,
                        ZigBeeBroadcastDestination.isBroadcast(destinationAddress));
                if (receivedCommand instanceof NetworkAddressResponse) {
                    logger.debug("Stefan sourceAddress={}",
                            ((NetworkAddressResponse) receivedCommand).getSourceAddress());
                }

                if (!ZigBeeBroadcastDestination.isBroadcast(destinationAddress)) {
                    logger.debug("Stefan: Not broadcast");
                    if (receivedCommand instanceof NetworkAddressResponse) {
                        if (!((ZdoRequest) command).getDestinationAddress()
                                .equals(((NetworkAddressResponse) receivedCommand).getSourceAddress())) {
                            logger.debug("Stefan: Addresses dont match");
                            // return false;
                        }
                    }
                }
                logger.debug("Stefan: Response matches: command={} response={}", command, receivedCommand);
                // return true;
            }

            if (Objects.equals(command.getTransactionId(), receivedCommand.getTransactionId())
                    && responseMatcher.isTransactionMatch(command, receivedCommand)) {
                // If the transaction state is TRANSMITTED, then we know that the transport layer
                // is providing state updates. We need to ensure in this case that we get the RX_ACK or RX_NAK
                // states so that we know that the transport layer has completed the APS transaction.
                // If the state is TRANSMITTED, then we have not yet received this, and we need to hold off
                // completing this transaction to avoid sending another transaction to the transport that it
                // Cannot currently handle.
                if (state == TransactionState.TRANSMITTED) {
                    state = TransactionState.RESPONDED;
                    completionCommand = receivedCommand;
                    logger.debug("Transaction response received - waiting TX_ACK: {}", this);
                } else {
                    completeTransaction(receivedCommand);
                }
            }
        }
    }
```

And *YES*, they are buggy, because I wanted to have them in `NetworkAddressRequest`, but somehow our logging was off in there. So I put it into the transaction and had to remove the `return` parts. But if one sees "Stefan: Addresses dont match", then the matcher below will device that the response does NOT match the request.

triller-telekom commented:
I have now added another commit where I fixed a bug that I had introduced (missed the "!" in the instance of check) and I am now testing this.

If everything is working as expected, I will remove the WIP and you can do a review.

cdjackson commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NetworkAddressRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/SimpleDescriptorRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLeaveRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/PowerDescriptorRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ActiveEndpointsRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UserDescriptorRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NodeDescriptorRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/BindRequest.java  4
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ComplexDescriptorRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementPermitJoiningRequest.java  4
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UnbindRequest.java  4
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  3
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementBindRequest.java  4
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/IeeeAddressRequest.java  3
         

Clones added
============
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/PowerDescriptorRequest.java  1
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ActiveEndpointsRequest.java  1
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/UserDescriptorRequest.java  1
- com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/NodeDescriptorRequest.java  1
         

Clones removed
==============
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementRoutingRequest.java  -1
+ com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zdo/command/ManagementLqiRequest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zsmartsystems/com-zsmartsystems-zigbee/pullRequest?prid=6068786&bid=19873994)

triller-telekom commented:
I have now tested it again with the devices from https://github.com/zsmartsystems/com.zsmartsystems.zigbee/issues/1114#issuecomment-680054006

and i was able to pair/unpair, re-pair and control them. So far I have not spotted any issues.

I left the 3 commits in the PR in case you want to understand the steps, but afterwards all of them should be squashed before merging them.

I would appreciate if you could give this a test. I am especially interested to see how it works with your 32 lamps setup.

triller-telekom commented:
To follow-up on this: We have tested it a bit further and have found no issues so far. The devices that we used in addition to mine were:

CLA60 RGBW OSRAM,
Bitron Video 902010/25,
Bitron Video AV2010/34
Bitron Home 902010/32

bitron-video-902010-21
bitron-video-902010-23
bitron-video-902010-24
bitron-video-902010-25
bitron-video-902010-28
bitron-video-902010-29
bitron-video-902010-29a
bitron-video-902010-32
bitron-video-902010-37
centralite-3200 (de)
centralite-3200 (gb)
centralite-3310
centralite-3400-D
osram-b40-dim-z3
paul-neuhaus-nlg-rgbw-light
paul-neuhaus-rgb-cct-light
sengled-Z01-A60EAE27

@cdjackson Did you already have a chance for testing this?

cdjackson commented:
Thanks Stefan. This sounds very promising. Unfortunately I've not had a chance to look at this (sorry). I will do so in the next couple of days though.

cdjackson commented:
Hey Stefan. Sorry for the slow response with this one... I gave this a test and it seems to be fine - I didn't really look at exactly what was happening with transactions, but with 32 bulbs network, from a "cold start" (ie no network state was persisted - I deleted all the data store XML files first) everything was discovered and came back up ok. So this seems to say that all transactions completed - there were some errors along the way, but the ones I saw were network layer (route errors) and probably just caused by the restart.

So, if you're happy we might as well merge this and get it into circulation.

triller-telekom commented:
Hey Chris,

sorry that I get back on this very late too, as I was on "vacation".

Your test with 32 bulbs sounds promising. And indeed, if the framework was able to discover 32 devices, a lot of (different) transactions had to be successful to achieve that.

I am fine with merging this. So please click the "merge" button :)

cdjackson commented:
Hey Stefan. No problem - I thought maybe you'd gone on holiday (at home? It's a strange time at the moment ;) ).

Thanks for this - I'll merge.
